sdk/java/src/main/java/sawtooth/sdk/processor/State.java

```diff
@@ -144,5 +149,59 @@ public State(Stream stream, String contextId) {
 
     return addressesThatWereSet;
   }
+  
+  /**Make an add event request on something happend on State Object for a specific context. 
+   * 
+   * @param eventType {String} - This is used to subscribe to events. It should
+   *     be globally unique and describe what, in general, has occurred
+   * @param attributes {Collection{key,value}} - Additional information about the event that
+   *     is transparent to the validator.  Attributes can be used by subscribers to
+   *     filter the type of events they receive   
+   * @param data {ByteString} - Additional information about the event that is
+   *     opaque to the validator 
+   * @throws InternalError something went wrong while processing transaction
+   */
 
+  public void addEvent(String eventType,
+                       Collection<java.util.Map.Entry<String, String>> 
+                       attributes,ByteString data)
+                       throws InternalError, InvalidTransactionException {    
+    if (attributes == null) {
+      attributes = new ArrayList<>();
+    }
+    sawtooth.sdk.protobuf.Event.Attribute.Builder attributeBuilder = null;
+    for (Map.Entry<String, String> entry : attributes) {
+      if (!entry.getKey().isEmpty()) {
+        attributeBuilder = sawtooth.sdk.protobuf.Event.Attribute.newBuilder()
+                           .setKey(entry.getKey())
+                           .setValue(entry.getValue());
+      }
+    }    
+    Event event = sawtooth.sdk.protobuf.Event.newBuilder()
+                  .setEventType(eventType)
+                  .setData(data)
+                  .addAttributes(attributeBuilder.build()).build();
+    TpEventAddRequest eventAddRequest = TpEventAddRequest
+                                        .newBuilder()
+                                        .setContextId(this.contextId)
+                                        .setEvent(event).build();
+    Future future = stream.send(MessageType.TP_EVENT_ADD_REQUEST, 
+                              eventAddRequest.toByteString());
+    TpEventAddResponse eventAddReponse = null; 
+    try {
+      eventAddReponse = TpEventAddResponse
+                        .parseFrom(future.getResult(TIME_OUT));
+    } catch (InterruptedException iee) {
+      throw new InternalError(iee.toString());
+    } catch (InvalidProtocolBufferException ipbe) {
+      throw new InternalError(ipbe.toString());
```

Comment:
That's probably not an internal error. Internal errors should be used with caution because the TPs will just keep doing the same thing until it passes.